File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/proxy/lab/proxy_foo.cpp
#include "proxy_foo.h"

proxy_foo::proxy_foo(real_foo* a_real_foo)
{
	set_my_foo(a_real_foo);
}

void proxy_foo::set_my_foo(real_foo* a_real_foo)
{
	if (!a_real_foo)
	{
		throw dp_exception("void proxy_foo::set_real_foo(real_foo * a_real_foo) -- a_real_foo == nullptr.");
	}
	my_foo = a_real_foo;
}

void proxy_foo::bar()
{
	std::cout << "before:" << std::endl;
	my_foo->bar();
	std::cout << "after:" << std::endl;

	std::cout << "before:" << std::endl;
	my_foo->bar();
	std::cout << "after:" << std::endl;

	std::cout << "before:" << std::endl;
	my_foo->bar();
	std::cout << "after:" << std::endl;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/proxy/solution/ProxyFoo.m
% ProxyFoo.m

classdef ProxyFoo < Foo
	properties (GetAccess = private, SetAccess = private)
		RealFoo % RealFoo
	end

	methods
		% constructor:
		function theProxyFoo = ProxyFoo(aRealFoo)
			theProxyFoo.RealFoo = aRealFoo;
		end

		function bar(theProxyFoo)
			fprintf('before:\n');

			theProxyFoo.RealFoo.bar;

			fprintf('after:\n');
		end

		function set.RealFoo(theProxyFoo, aRealFoo)
			if isempty(aRealFoo)
				error('There must be a non-empty RealFoo.');
			end

			theProxyFoo.RealFoo = aRealFoo;
		end
	end
end
